dmutils/email/dm_notify.py

```diff
@@ -0,0 +1,123 @@
+        return self._sent_references_cache
+
+    def _update_cache(self, reference):
+        """If the cache has been instantiated then cache the new reference."""
+        if self._sent_references_cache is not None:
+            self._sent_references_cache.append(reference)
+
+    def has_been_sent(self, reference):
+        """Checks for a matching reference in our list of delivered references."""
+        return reference in self.get_delivered_references()
+
+    @staticmethod
+    def get_reference(email_address, template_id, personalisation=None, **kwargs):
+        """
+        Method to return the standard reference given the variables the email is sent with.
+
+        :param email_address: Emails recipient
+        :param template_id: Emails template ID on Notify
+        :param personalisation: Template parameters
+        :param kwargs: Extra data passed to the reference eg. {'notes': 'Manual resend'}
+        :return: Hashed string 'reference' to be passed to client.send_email_notification or self.send_email
+        """
+        details_dict = OrderedDict([
+            ('email_address', email_address),
+            ('template_id', template_id),
+            ('personalisation', ','.join(stringify_dict(personalisation))),
+        ])
+        details_string = '|'.join(stringify_dict(details_dict))
+
+        return hash_string(details_string)
+
+    @staticmethod
+    def get_error_message(email_address, error):
+        """Format a logical error message from the error response."""
+        messages = []
+        message_prefix = 'Error sending message to {email_address}: '.format(email_address=email_address)
+        message_string = '{status_code} {error}: {message}'
+
+        for message in error.message:
+            format_kwargs = {
+                'status_code': error.status_code,
+                'error': message['error'],
+                'message': message['message'],
+            }
+            messages.append(message_string.format(**format_kwargs))
+        return message_prefix + ', '.join(messages)
+
+    def send_email(self, email_address, template_id, personalisation=None, no_resend=False):
+        """
+        Method to send an email using the Notify api.
+
+        :param email_address: String email address for recipient
+        :param template_id: Template accessible on the Notify account whose  api_key you instantiated the class with
+        :param personalisation: The template variables, dict
+        :param no_resend: Bool, instantiate the delivered reference cache and ensure we are not sending duplicates
+        :return: response from the api. For more information see https://github.com/alphagov/notifications-python-client
+        """
+        reference = self.get_reference(email_address, template_id, personalisation)
+        if no_resend and self.has_been_sent(reference):
+            self.logger.warn("Email {template_id} has already been sent to {email_address} with Notify", extra=dict(
```

Comment:
Hmmm, not sure about this one. We've taken the time to set the `resend=False` flag because we don't want to send duplicates. The code then tries to send a duplicate. I'd prefer visibility on that.
I'll change it if you feel strongly about it.